packages/node/src/indexer/api.service.ts

```diff
@@ -26,24 +34,117 @@ const NOT_SUPPORT = (name: string) => () => {
 
 // https://github.com/polkadot-js/api/blob/12750bc83d8d7f01957896a80a7ba948ba3690b7/packages/rpc-provider/src/ws/index.ts#L43
 const RETRY_DELAY = 2_500;
+const MAX_RECONNECT_ATTEMPTS = 5;
 
 const logger = getLogger('api');
 
+export class ApiResponseTimeBuffer extends Queue<number> {
+  constructor(capacity: number) {
+    super(capacity);
+  }
+
+  average() {
+    if (this.items.length === 0) {
+      return 0;
+    }
+    return this.items.reduce((a, b) => a + b, 0) / this.items.length;
+  }
+}
+
+export class ApiLoadBalancer {
+  private buffers: Record<number, ApiResponseTimeBuffer> = {};
+
+  constructor(numberOfConnections: number, batchCapacity: number) {
+    // fill buffers with ApiResponseTimeBuffer object with capacity of batchCapacity
+    range(0, numberOfConnections).forEach((n) => {
+      this.buffers[n] = new ApiResponseTimeBuffer(batchCapacity);
+    });
+    logger.info(`length: ${Object.keys(this.buffers).length}`);
+  }
+
+  addToBuffer(connectionIndex: number, responseTime: number) {
+    if (
+      this.buffers[connectionIndex].size ===
+      this.buffers[connectionIndex].capacity
+    ) {
+      this.buffers[connectionIndex].take();
+    }
+    this.buffers[connectionIndex].put(responseTime);
+  }
+
+  getWeights(disconnectedIndices?: number[]): Record<number, number> {
+    const weightsMap: Record<number, number> = {};
+    Object.keys(this.buffers).map((key) => {
+      if (disconnectedIndices?.includes(toNumber(key))) {
+        return;
+      }
+      weightsMap[key] = this.buffers[key].average();
+    });
+    if (Object.keys(weightsMap).length === 0) {
+      return {};
+    }
+
+    const total = Object.values(weightsMap).reduce((a, b) => a + b, 0);
+    //deal with the case where average is 0
+    if (total === 0) {
+      Object.keys(weightsMap).map((key) => {
+        weightsMap[key] = 1 / Object.keys(weightsMap).length;
+      });
+      return weightsMap;
+    }
+    Object.keys(weightsMap).map((key) => {
+      weightsMap[key] = weightsMap[key] / total;
+    });
+    return weightsMap;
+  }
+}
+
 @Injectable()
 export class ApiService implements OnApplicationShutdown {
-  private api: ApiPromise;
+  private connectionPool: Record<number, ApiPromise> = {};
+  private disconnectedApis: Record<number, ApiPromise> = {};
```

Comment:
I don't think we need to keep track of this. We should only need to keep reference to all and connected apis